tortoisebeacon/tortoisebeacon.go

```diff
@@ -0,0 +1,652 @@
+		epochATXGetter:       epochATXGetter,
+		tortoiseBeaconDB:     tortoiseBeaconDB,
+		weakCoin:             weakCoin,
+		layerTicker:          layerTicker,
+		networkDelta:         time.Duration(conf.WakeupDelta) * time.Second,
+		currentRounds:        make(map[types.EpochID]int),
+		timelyProposalsMap:   make(map[types.Hash32][]types.ATXID),
+		timelyProposalsList:  make(map[types.EpochID][][]types.ATXID),
+		delayedProposalsMap:  make(map[types.Hash32][]types.ATXID),
+		delayedProposalsList: make(map[types.EpochID][][]types.ATXID),
+		lateProposalsList:    make(map[types.EpochID][][]types.ATXID),
+		votesFor:             make(map[EpochRoundPair][]types.Hash32),
+		votesAgainst:         make(map[EpochRoundPair][]types.Hash32),
+		beacons:              make(map[types.EpochID]types.Hash32),
+		beaconsReady:         make(map[types.EpochID]chan struct{}),
+	}
+}
+
+// Start starts listening for layers and outputs.
+func (tb *TortoiseBeacon) Start() error {
+	tb.Log.Info("Starting %v", protoName)
+
+	tb.backgroundWG.Add(1)
+	go func() {
+		defer tb.backgroundWG.Done()
+
+		tb.listenLayers()
+	}()
+
+	tb.backgroundWG.Add(1)
+	go func() {
+		defer tb.backgroundWG.Done()
+
+		if err := tb.listenMessages(); err != nil {
+			tb.Log.With().Error("Error while listening for messages: %v", log.Err(err))
+			return
+		}
+	}()
+
+	tb.backgroundWG.Add(1)
+	go func() {
+		defer tb.backgroundWG.Done()
+
+		tb.cleanupLoop()
+	}()
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) Close() error {
+	tb.Log.Info("Closing %v", protoName)
+	tb.Closer.Close()
+	tb.backgroundWG.Wait() // Wait until background goroutines finish
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) Get(epochID types.EpochID) (types.Hash32, error) {
+	if tb.tortoiseBeaconDB != nil {
+		if val, ok := tb.tortoiseBeaconDB.GetTortoiseBeacon(epochID); ok {
+			return val, nil
+		}
+	}
+
+	tb.beaconsMu.RLock()
+	beacon, ok := tb.beacons[epochID]
+	tb.beaconsMu.RUnlock()
+
+	if !ok {
+		return types.Hash32{}, ErrBeaconNotCalculated
+	}
+
+	if tb.tortoiseBeaconDB != nil {
+		if err := tb.tortoiseBeaconDB.SetTortoiseBeacon(epochID, beacon); err != nil {
+			return types.Hash32{}, err
+		}
+	}
+
+	return beacon, nil
+}
+
+// Wait waits until beacon for this epoch becomes ready.
+func (tb *TortoiseBeacon) Wait(epochID types.EpochID) error {
+	if tb.tortoiseBeaconDB != nil {
+		if _, ok := tb.tortoiseBeaconDB.GetTortoiseBeacon(epochID); ok {
+			return nil
+		}
+	}
+
+	tb.beaconsMu.RLock()
+	ch, ok := tb.beaconsReady[epochID]
+	tb.beaconsMu.RUnlock()
+
+	if !ok {
+		return ErrBeaconNotCalculated
+	}
+
+	<-ch
+	return nil
+}
+
+func (tb *TortoiseBeacon) cleanupLoop() {
+	ticker := time.NewTicker(cleanupInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-tb.CloseChannel():
+			return
+		case <-ticker.C:
+			tb.cleanup()
+		}
+	}
+}
+
+func (tb *TortoiseBeacon) cleanup() {
+	// TODO(nkryuchkov): implement a better solution, consider https://github.com/golang/go/issues/20135
```

Comment:
also cleaning of beacons maps must be done once in a while and not just when closing, since it will run for a long time and eventually contain all epoch beacons